Revamped add fields::dft_fields_norm2

smartalecH opened the pull request:
Fixes #1711

Similar to #1718, except it does an L2 norm over the already calculated DFT fields, rather than manually recomputing (a modified version of) them. This means the user doesn't have to run this as often (before you had to "sample" fast enough to prevent aliasing). Plus there are fewer FLOPs with this approach.

I'm sure there's a reason @stevengj didn't do things this way, just not sure what that may be.

It passed the adjoint solver test suite locally. I reran the first tutorial and saw similar relative errors in the gradients as the old method.

We should probably do some profiling to see what convergence thresholds are really needed for "accurate" gradients.

smartalecH commented on python/adjoint/optimization_problem.py:
```diff
@@ -567,44 +535,9 @@ def _stop(sim):
-            relative_change = np.mean(
-                relative_change)  # average across monitors
+            current_fields  = simob.fields.dft_fields_norm()
+            relative_change =  np.abs(previous_fields-current_fields) /  previous_fields   
```
Fixed.

smartalecH commented on src/dft.cpp:
```diff
@@ -265,6 +265,64 @@ void dft_chunk::update_dft(double time) {
+  return sum;
+}
+
+// return the minimum abs(freq) over all DFT chunks
```
Fixed.

smartalecH commented:
`stop_when_dft_decayed` now supports *both* methods for computing the decay rate (see the `method` flag). Turns out both methods now give very similar results with the adjoint solver test. The default is the more conservative method in #1718.

The decay rate is calculated every `1/sim.fields.dft_maxfreq()` timesteps for both methods, which means it is called *much* more frequently than before. As a result, the decay tolerance needs to b `<1e-10` for the adjoint tests to pass. This doesn't mean the simulation is running longer than before -- quite the opposite actually. Because it's updating so frequently, it can shut off as soon as that tolerance is hit. Before, the update was only called every 50 timesteps (a rather large default value).

The `OptimizationProblem` no longer uses the `decay_dt` or `decay_fields` parameters, but I left them in to not break (mostly my) existing code. I'm sure with future releases we can drop those parameters.

stevengj commented:
In principle, we should set how often we check this to be `max(1/(maxfreq*dt), decimation factors)` so that we never halt without updating the DFTs first.

stevengj commented:
Since they are performing similarly, maybe we should just do the less conservative approach (your initial approach).

smartalecH commented:
> In principle, we should set how often we check this to be max(1/(maxfreq*dt), decimation factors) so that we never halt without updating the DFTs first.

Fixed.  I added a function, `min_decimation()`, that pulls the minimum decimation factor from all the DFT monitors (very similar to `dft_maxfreq()`.

> Since they are performing similarly, maybe we should just do the less conservative approach (your initial approach), and just call it dft_norm2 or even just norm2.

Fixed.

smartalecH commented:
I readded the original adjoint solver test file, `test_adjoint_solver.py`. It seemed to have been deleted in #1749.

stevengj commented on src/dft.cpp:
```diff
@@ -283,6 +283,81 @@ void dft_chunk::update_dft(double time) {
+/* Return the L2 norm of the DFTs themselves.  This is useful
+   to check whether the simulation is finished (whether all relevant fields have decayed).
+   (Collective operation.) */
+double fields::dft_norm2() {
```
I would just call it `dft_norm` since it is not squared (unlike `fields_chunk::dft_norm2` or `dft_chunk::norm2`)

smartalecH replied:
Fixed.

stevengj commented on python/simulation.py:
```diff
@@ -4425,6 +4425,41 @@ def _stop(sim):
+    closure = {'previous_fields':1, 't0':0, 'dt':0}
+    def _stop(_sim):
+        if _sim.fields.t == 0:
+            closure['dt'] = max(int(1/_sim.fields.dft_maxfreq()/_sim.fields.dt),int(_sim.fields.min_decimation()))
```
Why is `int(...)` required for `min_decimation`?  Isn't that already returning an integer?

stevengj replied:
Anyway, you can probably remove the `int` calls here completely since there's no reason `dt` needs to be an integer for the `t ≤ dt + t0` check.

smartalecH replied:
Fixed.

stevengj commented on python/simulation.py:
```diff
@@ -4425,6 +4425,41 @@ def _stop(sim):
+        else:
+            previous_fields = closure['previous_fields']
+            current_fields  = _sim.fields.dft_norm2()
+            relative_change = np.abs(previous_fields-current_fields) /  previous_fields   
```
I would `return false if previous_fields == 0`, and then initialize `previous_fields` to `0`.

smartalecH replied:
Fixed.

stevengj commented on python/simulation.py:
```diff
@@ -4425,6 +4425,41 @@ def _stop(sim):
+            if mp.verbosity > 0:
+                fmt = "DFT decay(t = {0:1.1f}): {1:0.4e}"
+                print(fmt.format(_sim.meep_time(), np.real(relative_change)))
+            return relative_change <= decay_by and _sim.round_time(
```
I'm not completely happy with the fact that the `decay_by` tolerance implicitly depends on `closure['dt']`.

(In contrast, my alternate proposal of `|fields| ≤ tol * max|fields|` has a `tol` that is independent of the sampling frequency.)

One alternative would be to check `change <= tol * maxchange`, where `change = np.abs(previous_fields-current_fields)` and `maxchange` is the maximum observed change, so that `tol` becomes independent of the sampling frequency.

smartalecH replied:
Why does `decay_by` implicitly depend on `closure['dt']`? We are only looking at a predetermined set of frequencies. So long as the frequencies are below the Nyquist limit, the sample rate shouldn't affect the decay tolerance, right?

> One alternative would be to check change <= tol * maxchange, where change = np.abs(previous_fields-current_fields) and maxchange is the maximum observed change over the simulation, so that tol becomes independent of the sampling frequency.

Aren't we already normalizing out any dependence? How different is the current implementation:
```relative_change = np.abs(previous_fields-current_fields) /  previous_fields ```

from your proposal:
```change = np.abs(previous_fields-current_fields) /  maxchange ```

The issue with `maxchange` is now you have to compute a Δ over every spatial point, which requires *caching* all of the dft fields.

stevengj replied:
No, by `maxchange` I mean the maximum value of `change` observed over all the previous sampling *times*.

smartalecH replied:
Fixed.

oskooi commented on python/simulation.py:
```diff
@@ -4425,6 +4425,41 @@ def _stop(sim):
+        if _sim.fields.t <= closure['dt'] + closure['t0']:
+            return False
+        elif maximum_run_time and _sim.round_time() > maximum_run_time:
+            return True
```
Shouldn't the second if-branch which checks if the `maximum_run_time` has been exceeded come *before* the first if-branch which checks that a time interval of `dt` has elapsed?

smartalecH replied:
Fixed.

oskooi commented on python/adjoint/optimization_problem.py:
```diff
@@ -130,7 +130,7 @@ def __init__(
 
         self.decay_by = decay_by
-        self.decay_fields = decay_fields
+        self.decay_fields = decay_fields # left for legacy, but doesn't do anything
```
Probably better to just remove this line here and also from `test_adjoint_solver.py` where it is also used?

smartalecH replied:
Fixed.

smartalecH commented:
Should be ready to merge once all tests pass.

stevengj commented on python/simulation.py:
```diff
@@ -4425,6 +4425,44 @@ def _stop(sim):
+            return False
+        else:
+            previous_fields = closure['previous_fields']
+            current_fields  = _sim.fields.dft_norm2()
```
```suggestion
            current_fields  = _sim.fields.dft_norm()
```

stevengj commented on python/simulation.py:
```diff
@@ -4425,6 +4425,44 @@ def _stop(sim):
+    run time (`maximum_run_time`).
+    """
+    # Record data in closure so that we can persitently edit
+    closure = {'previous_fields':0, 't0':0, 'dt':0, 'maxchange':1}
```
```suggestion
    closure = {'previous_fields':0, 't0':0, 'dt':0, 'maxchange':0}
```

stevengj commented on python/simulation.py:
```diff
@@ -4425,6 +4425,44 @@ def _stop(sim):
+            if mp.verbosity > 0:
+                fmt = "DFT decay(t = {0:1.1f}): {1:0.4e}"
+                print(fmt.format(_sim.meep_time(), np.real(change/closure['maxchange'])))
+            return (change/closure['maxchange']) <= tol and _sim.round_time() >= minimum_run_time
```
```suggestion
            return (change <= closure['maxchange'] * tol and _sim.round_time() >= minimum_run_time
```

stevengj replied:
(to handle case where `maxchange == 0`)

smartalecH commented:
Finally passing tests and ready to merge

oskooi commented on python/adjoint/optimization_problem.py:
```diff
@@ -69,12 +69,10 @@ def __init__(
-        decay_dt=50,
-        decay_fields=[mp.Ez],
-        decay_by=1e-6,
+        decay_by=1e-11,
```
Why does the default tolerance need to be reduced by five orders of magnitude? Won't such a small default value potentially lead to longer runtimes than are actually necessary particularly since the discretization errors are likely several orders of magnitude larger?

stevengj replied:
The default tolerance now has a completely different meaning than before, so I'm not sure whether this actually corresponds to an increase in runtime…

smartalecH replied:
Exactly. This PR actually *reduces* the runtime from `master`, since it's able to check so frequently. But the tolerance corresponds to a completely different convergence check, which is why it's value is so different than before. 

Essentially the default tolerance is chosen to make the tests pass. But the runtime is still minimized (and less than `master`).
